src/coreclr/jit/lsrariscv64.cpp

```diff
@@ -286,10 +286,65 @@ int LinearScan::BuildNode(GenTree* tree)
         case GT_MOD:
         case GT_UMOD:
         case GT_DIV:
-        case GT_MULHI:
         case GT_UDIV:
         {
             srcCount = BuildBinaryUses(tree->AsOp());
+
+            GenTree* divisorOp = tree->gtGetOp2();
+            if (!varTypeIsFloating(tree->TypeGet()) &&
+                !(divisorOp->IsIntegralConst(0) || divisorOp->GetRegNum() == REG_R0))
+            {
+                bool needTemp = false;
+                if (divisorOp->isContainedIntOrIImmed())
+                {
+                    needTemp = true; // divisorReg
+                }
+                else if (tree->OperIsCommutative())
+                {
+                    if (tree->gtGetOp1()->isContainedIntOrIImmed())
+                        needTemp = true; // reg1
+                }
+
+                if (tree->gtOper == GT_DIV || tree->gtOper == GT_MOD)
```

Comment:
Suppose that at this point we know we need temp. Does it still make sense to enter this branch?

Reply:
> Suppose that at this point we know we need temp

Do we? needTemp is true only when one of the ifs above is entered.

Reply:
What I'm trying to say is: I guess that if needTemp is true we don't to enter this block at all.

Reply:
ok, now I get it:) yes, you're right